From 0a8021958ea816fcc033da01ed5a4d882f39086c Mon Sep 17 00:00:00 2001
From: Hongxu Jia <hongxu.jia@windriver.com>
Date: Sat, 29 Jun 2019 14:46:57 +0800
Subject: [PATCH] support authentication for kickstart

We add option 'ksuser' and 'kspasswd' in /proc/cmdline to support
authentication while downloading kickstart from web server.

Upstream-Status: Inappropriate [oe specific]

Signed-off-by: Hongxu Jia <hongxu.jia@windriver.com>

Update patch for anaconda 32:
* replace self.cmdline with kernel_arguments
* update patch context

Signed-off-by: Kai Kang <kai.kang@windriver.com>

Rebase for anaconda 34:

Signed-off-by: Kai Kang <kai.kang@windriver.com>

It requires option prefix 'inst.' in anaconda 34. Update the error
message.

Signed-off-by: Kai Kang <kai.kang@windriver.com>

Rebase for anaconda 37.

Signed-off-by: Kai Kang <kai.kang@windriver.com>

Rebase for anaconda 42 on 20250328.

Signed-off-by: Kai Kang <kai.kang@windriver.com>
---
 pyanaconda/flags.py                                |  6 ++++
 pyanaconda/kickstart.py                            | 37 +++++++++++++++++++---
 .../modules/boss/kickstart_manager/parser.py       | 10 +++---
 3 files changed, 43 insertions(+), 10 deletions(-)

diff --git a/pyanaconda/flags.py b/pyanaconda/flags.py
index d52fa2a7c4..28ffac7669 100644
--- a/pyanaconda/flags.py
+++ b/pyanaconda/flags.py
@@ -18,6 +18,7 @@
 #
 from pyanaconda.anaconda_loggers import get_module_logger
 from pyanaconda.core.constants import ANACONDA_ENVIRON
+from pyanaconda.core.kernel import kernel_arguments
 
 log = get_module_logger(__name__)
 
@@ -45,6 +46,11 @@ class Flags:
         self.kexec = False
         # current runtime environments
         self.environs = [ANACONDA_ENVIRON]
+
+        # Authentication for downloading kickstart file through http
+        self.ksuser = kernel_arguments.get('ksuser')
+        self.kspasswd = kernel_arguments.get('kspasswd')
+
         # Lock it down: no more creating new flags!
         self.__dict__['_in_init'] = False
 
diff --git a/pyanaconda/kickstart.py b/pyanaconda/kickstart.py
index 70068f41f8..1172ad8f9d 100644
--- a/pyanaconda/kickstart.py
+++ b/pyanaconda/kickstart.py
@@ -26,7 +26,7 @@ from contextlib import contextmanager
 
 from pykickstart.base import KickstartCommand, RemovedCommand
 from pykickstart.constants import KS_SCRIPT_PRE
-from pykickstart.errors import KickstartError, KickstartParseWarning
+from pykickstart.errors import KickstartError, KickstartParseWarning, KickstartAuthError
 from pykickstart.ko import KickstartObject
 from pykickstart.parser import KickstartParser
 from pykickstart.parser import Script as KSScript
@@ -225,8 +225,35 @@ class AnacondaKSHandler(superclass):
         return super().__str__() + "\n" + modules
 
 
-class AnacondaPreParser(KickstartParser):
-    # A subclass of KickstartParser that only looks for %pre scripts and
+class OEKickstartParser(KickstartParser):
+    # A subclass of KickstartParser that parse the authentication failure
+    # while downloading kickstart file from web server
+    def __init__(self, handler, followIncludes=True, errorsAreFatal=True,
+                  missingIncludeIsFatal=True):
+        super().__init__(handler, missingIncludeIsFatal=False)
+        self.username = flags.ksuser
+        self.password = flags.kspasswd
+
+    def readKickstart(self, f, reset=True):
+        while True:
+            try:
+                super().readKickstart(f, reset=reset,
+                                              username=self.username,
+                                              password=self.password)
+
+                # Make sure the following readKickstart invoking will success
+                flags.ksuser = self.username
+                flags.kspasswd = self.password
+                break
+            except KickstartAuthError as e:
+                msg = 'Append "inst.ksuser=<username> inst.kspasswd=<password>" to boot command'
+                raise KickstartError(msg)
+            except:
+                raise
+
+
+class AnacondaPreParser(OEKickstartParser):
+    # A subclass of OEKickstartParser that only looks for %pre scripts and
     # sets them up to be run.  All other scripts and commands are ignored.
     def __init__(self, handler):
         super().__init__(handler, missingIncludeIsFatal=False)
@@ -245,7 +272,7 @@ class AnacondaPreParser(KickstartParser):
         self.registerSection(NullSection(self.handler, sectionOpen="%certificate"))
 
 
-class AnacondaKSParser(KickstartParser):
+class AnacondaKSParser(OEKickstartParser):
     def __init__(self, handler, scriptClass=AnacondaKSScript):
         self.scriptClass = scriptClass
         super().__init__(handler)
@@ -254,7 +281,7 @@ class AnacondaKSParser(KickstartParser):
         if not self.handler:
             return
 
-        return KickstartParser.handleCommand(self, lineno, args)
+        return super().handleCommand(lineno, args)
 
     def setupSections(self):
         self.registerSection(PreScriptSection(self.handler, dataObj=self.scriptClass))
diff --git a/pyanaconda/modules/boss/kickstart_manager/parser.py b/pyanaconda/modules/boss/kickstart_manager/parser.py
index 97644e7b46..7cdc00d86e 100644
--- a/pyanaconda/modules/boss/kickstart_manager/parser.py
+++ b/pyanaconda/modules/boss/kickstart_manager/parser.py
@@ -18,7 +18,7 @@
 
 __all__ = ["VALID_SECTIONS_ANACONDA", "SplitKickstartParser"]
 
-from pykickstart.parser import KickstartParser
+from pyanaconda.kickstart import OEKickstartParser
 from pykickstart.sections import Section
 
 from pyanaconda.modules.boss.kickstart_manager.element import (
@@ -84,7 +84,7 @@ class StoreSection(Section):
         self.lines = []
 
 
-class SplitKickstartParser(KickstartParser):
+class SplitKickstartParser(OEKickstartParser):
     """Kickstart parser for storing kickstart elements.
 
     Stores kickstart elements (commands, sections, addons) with their line
@@ -135,9 +135,9 @@ class SplitKickstartParser(KickstartParser):
                  kickstart files
         :rtype: KickstartElements
         """
-        with open(filename, "r") as f:
-            kickstart = f.read()
-        return self.split_from_string(kickstart, filename=filename)
+        self._current_ks_filename = filename or self.unknown_filename
+        self.readKickstart(filename)
+        return self._result
 
     def split_from_string(self, kickstart, filename=None):
         """Split the kickstart given as string into elements.
-- 
2.7.4

